homeassistant/components/tuya/cover.py

```diff
@@ -50,7 +50,7 @@ class TuyaCoverEntityDescription(CoverEntityDescription):
             key=DPCode.CONTROL,
             name="Curtain",
             current_state=DPCode.SITUATION_SET,
-            current_position=DPCode.PERCENT_STATE,
+            current_position=DPCode.PERCENT_CONTROL,
```

Comment:
This is actually changing something documented:

https://developer.tuya.com/en/docs/iot/s?id=K9gf48qy7wkre#title-1-Support%20reporting%20instructions%20to%20the%20cloud%20only

![image](https://user-images.githubusercontent.com/195327/141698993-9ee06c10-2819-4c5b-a709-72b97d3316b2.png)

So this change will be breaking it for others.

Reply:
This is actually the same concern as I raised earlier in this comment: https://github.com/home-assistant/core/pull/59676#discussion_r748871559

Reply:
from the value that i here see on the curtain, the percent_state value is keep getting 0 value, we reverse the value so we always get 100. That the reason why the position is keep 100.
I agree that it  could break other stuffs.
I've also check on the integration before on tuya V2:
https://github.com/adamd9/tuya-home-assistant/blob/83a0bf9002d497bf0e564377feab73bdde15a374/custom_components/tuya_v2/cover.py#L30
it is also using "PERCENT_CONTROL" instead on "PERCENT_STATE"

Reply:
screenshot with working state:
![image](https://user-images.githubusercontent.com/1487966/141702706-bf553e89-d3bb-4f08-9eab-9d321db50bc6.png)

Reply:
This shows that different devices do different things for some reason. I for instance don't get the PERCENT_STATE attribute. Only the PERCENT_CONTROL one. Changing "current_state=DPCode. SITUATION_SET" to "current_state=DPCode.PERCENT_CONTROL" and "current_position=DPCode.PERCENT_STATE" to "current_position=DPCode.PERCENT_CONTROL" would fix all my issues for example. Don't have a clue how to continue with that ...

Reply:
@leeyuentuen
> Value from tuya IoT

Sure, but the current code I've build using other devices, so this change will break those. This PR needs to be updated to support both cases, we could prefer using `DPCode.PERCENT_CONTROL` and fallback to `DPCode.PERCENT_STATE`.

@acmaarts

>  would fix all my issues for example. Don't have a clue how to continue with that ...

Could you please stick with code reviewing in PRs? This is not the place for speculation or questions on how to contribute. Thanks 👍

Reply:
@frenck Despite it is documented that way, It does not work that way for some reason. The values listed as "Support reporting instructions to the cloud only" are not returned in my case. All the others are.

Reply:
it is indeed a little bite strange, but I came from tuya V2 where the curtain works before.

Reply:
> it is indeed a little bit strange, but I came from tuya V2 where the curtain works before.

We don't care about Tuya V2, that was a custom integration that had its own set of issues and problems. We are working on the core integration here.

I'm saying that the current code is based on a device I had for testing, this change breaks it in order to fix yours. That is not a solution, we need to support both. This PR needs to adjust to that.

Reply:
I've changed something that keeps the current DPCode, but it will try to check the work_state.

Reply:
?!

I'm lost now... why did you do that? And what has that to do with this comment?

The issue you presented is that the wrong DP code is used for determining the percentage, at this point, where is now?!

Reply:
I've kept the current value to keep the current device working.

In the function "current_cover_position", we try to get the current- and set-position.
I've put in the code to check if we can use the set_position or using the current_position

so if we are closing the curtain and our current_position is something bigger than 0 then we expect that we can use the current_position dpcode. otherwise, we change them to use set_position (because the current_position is not used in my case here)

when we are opening the curtain and the current position is 0, then we know that the current_positiion value is not used. so in this case we also switch to use set_position dpcode.

In this case, we don't change the global code of the current  curtains devices if they are still using the 'old' current_position variable

To determine the correct percentage with the dpcode, it is at line (using the work-stage and based if it is open or close) 
https://github.com/home-assistant/core/blob/e7f3ee5606f2ddf474627b66c2624b799acc6339/homeassistant/components/tuya/cover.py#L255
and on line 
https://github.com/home-assistant/core/blob/e7f3ee5606f2ddf474627b66c2624b799acc6339/homeassistant/components/tuya/cover.py#L258

Reply:
> I've kept the current value to keep the current device working.

But? We should use both?

See my previous comment here: <https://github.com/home-assistant/core/pull/59676#discussion_r749189555>

> This PR needs to be updated to support both cases, we could prefer using DPCode.PERCENT_CONTROL and fallback to DPCode.PERCENT_STATE.

That is what needs to happen. Work state isn't related to this all, so I'm not sure how this works in your PR now and why it was even added?

Reply:
ah ok, i'll take a look on it.

Reply:
@frenck I've not done a rework, can you take a look at it?